File: System-Design/Caching/LFU-Cache/lfuCache.ts
class LFUNode {
    public freq: number = 1;
    public prev: LFUNode | null = null;
    public next: LFUNode | null = null;

    constructor(public key: number, public val: number) {}
}

class DoublyLinkedList {
    public head: LFUNode;
    public tail: LFUNode;
    public size: number = 0;

    constructor() {
        this.head = new LFUNode(0, 0);
        this.tail = new LFUNode(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    public insertHead(node: LFUNode): void {
        const nxt = this.head.next!;
        this.head.next = node;
        node.prev = this.head;
        node.next = nxt;
        nxt.prev = node;
        this.size++;
    }

    public remove(node: LFUNode): void {
        const prev = node.prev!;
        const nxt = node.next!;
        prev.next = nxt;
        nxt.prev = prev;
        this.size--;
    }

    public popTail(): LFUNode | null {
        if (this.size > 0) {
            const tailNode = this.tail.prev!;
            this.remove(tailNode);
            return tailNode;
        }
        return null;
    }
}

export class LFUCache {
    private capacity: number;
    private minFreq: number = 0;
    private keyToNode: Map<number, LFUNode> = new Map();
    private freqToList: Map<number, DoublyLinkedList> = new Map();

    constructor(capacity: number) {
        this.capacity = capacity;
    }

    private updateFreq(node: LFUNode): void {
        const oldFreq = node.freq;
        this.freqToList.get(oldFreq)!.remove(node);

        if (oldFreq === this.minFreq && this.freqToList.get(oldFreq)!.size === 0) {
            this.minFreq++;
        }

        node.freq++;
        const newFreq = node.freq;
        
        if (!this.freqToList.has(newFreq)) {
            this.freqToList.set(newFreq, new DoublyLinkedList());
        }
        this.freqToList.get(newFreq)!.insertHead(node);
    }

    public get(key: number): number {
        if (!this.keyToNode.has(key)) return -1;
        
        const node = this.keyToNode.get(key)!;
        this.updateFreq(node);
        return node.val;
    }

    public put(key: number, value: number): void {
        if (this.capacity === 0) return;

        if (this.keyToNode.has(key)) {
            const node = this.keyToNode.get(key)!;
            node.val = value;
            this.updateFreq(node);
            return;
        }

        if (this.keyToNode.size >= this.capacity) {
            const lruNode = this.freqToList.get(this.minFreq)!.popTail()!;
            this.keyToNode.delete(lruNode.key);
        }

        const newNode = new LFUNode(key, value);
        this.keyToNode.set(key, newNode);
        this.minFreq = 1;

        if (!this.freqToList.has(1)) {
            this.freqToList.set(1, new DoublyLinkedList());
        }
        this.freqToList.get(1)!.insertHead(newNode);
    }
}

// --- CI/CD Automated Test ---
const lfu = new LFUCache(2);

lfu.put(1, 1);
lfu.put(2, 2);

const p1 = lfu.get(1) === 1;

lfu.put(3, 3); // Evicts 2
const p2 = lfu.get(2) === -1;
const p3 = lfu.get(3) === 3;

lfu.put(4, 4); // Evicts 1
const p4 = lfu.get(1) === -1;
const p5 = lfu.get(3) === 3;
const p6 = lfu.get(4) === 4;

if (p1 && p2 && p3 && p4 && p5 && p6) {
    console.log("TypeScript LFU Cache Test Passed!");
} else {
    process.exit(1);
}

File: System-Design/Caching/LFU-Cache/lfu_cache.cpp
#include <iostream>
#include <unordered_map>
#include <cassert>

using namespace std;

struct Node {
    int key, val, freq;
    Node* prev;
    Node* next;
    Node(int k, int v) : key(k), val(v), freq(1), prev(nullptr), next(nullptr) {}
};

class DoublyLinkedList {
public:
    Node* head;
    Node* tail;
    int size;

    DoublyLinkedList() {
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
        size = 0;
    }

    ~DoublyLinkedList() {
        Node* curr = head;
        while (curr != nullptr) {
            Node* nxt = curr->next;
            delete curr;
            curr = nxt;
        }
    }

    void insertHead(Node* node) {
        Node* nxt = head->next;
        head->next = node;
        node->prev = head;
        node->next = nxt;
        nxt->prev = node;
        size++;
    }

    void remove(Node* node) {
        Node* prev = node->prev;
        Node* nxt = node->next;
        prev->next = nxt;
        nxt->prev = prev;
        size--;
    }

    Node* popTail() {
        if (size > 0) {
            Node* tailNode = tail->prev;
            remove(tailNode);
            return tailNode;
        }
        return nullptr;
    }
};

class LFUCache {
private:
    int capacity;
    int minFreq;
    unordered_map<int, Node*> keyToNode;
    unordered_map<int, DoublyLinkedList*> freqToList;

    void updateFreq(Node* node) {
        int oldFreq = node->freq;
        freqToList[oldFreq]->remove(node);

        if (oldFreq == minFreq && freqToList[oldFreq]->size == 0) {
            minFreq++;
        }

        node->freq++;
        int newFreq = node->freq;

        if (freqToList.find(newFreq) == freqToList.end()) {
            freqToList[newFreq] = new DoublyLinkedList();
        }
        freqToList[newFreq]->insertHead(node);
    }

public:
    LFUCache(int cap) : capacity(cap), minFreq(0) {}

    ~LFUCache() {
        for (auto& pair : freqToList) {
            delete pair.second;
        }
        // Nodes inside are deleted by DoublyLinkedList destructor
    }

    int get(int key) {
        if (keyToNode.find(key) == keyToNode.end()) return -1;
        
        Node* node = keyToNode[key];
        updateFreq(node);
        return node->val;
    }

    void put(int key, int value) {
        if (capacity == 0) return;

        if (keyToNode.find(key) != keyToNode.end()) {
            Node* node = keyToNode[key];
            node->val = value;
            updateFreq(node);
            return;
        }

        if (keyToNode.size() >= capacity) {
            Node* lruNode = freqToList[minFreq]->popTail();
            keyToNode.erase(lruNode->key);
            delete lruNode;
        }

        Node* newNode = new Node(key, value);
        keyToNode[key] = newNode;
        minFreq = 1;

        if (freqToList.find(1) == freqToList.end()) {
            freqToList[1] = new DoublyLinkedList();
        }
        freqToList[1]->insertHead(newNode);
    }
};

// --- CI/CD Automated Test ---
int main() {
    LFUCache lfu(2);

    lfu.put(1, 1);
    lfu.put(2, 2);

    assert(lfu.get(1) == 1);

    lfu.put(3, 3); // Evicts 2
    assert(lfu.get(2) == -1);
    assert(lfu.get(3) == 3);

    lfu.put(4, 4); // Evicts 1
    assert(lfu.get(1) == -1);
    assert(lfu.get(3) == 3);
    assert(lfu.get(4) == 4);

    cout << "C++ LFU Cache Test Passed!\n";
    return 0;
}